nussl/core/effects.py

```diff
@@ -0,0 +1,584 @@
+or a python-ffmpeg stream, depending on the specific effect. To apply the effect on an AudioSignal, 
+build_effects_sox or build_effects_ffmpeg must be called with the AudioSignal, and a list of
+SoXFilters or FFmpegFilters, respectively. 
+
+>>> import nussl.effects
+>>> tremolo_filter = effects.tremolo(5, .7)
+>>> new_signal = effects.build_effects_ffmpeg(audio_signal, [tremolo_filter])
+
+Because of this schema, and the requirement that caller must know which effect is a SoX effect
+or a FFmpeg effect, we recommend that all users use the AudioSignal hooks for applying effects
+rather than calling any functions in this library. 
+
+This line is equivalent to the above snippet
+
+>>> new_signal = audio_signal.tremolo(5, .7).build_effects()
+
+"""
+
+class FilterFunction:
+    """
+    The FilterFunction class is an abstract class for functions that take 
+    audio processing streams, such as ffmpeg-python and pysndfx
+    """
+
+    def __call__(self, stream):
+        return self.func(stream)
+
+    def func(self, stream):
+        pass
+
+
+class FFmpegFilter(FilterFunction):
+    """
+    FFmpegFilter is an object returned by FFmpeg effects in effects.py
+    To use them, build_effects_ffmpeg can take a list of effects, and apply them onto an 
+    AudioSignal object. 
+    """
+    def __init__(self, _filter, **filter_kwargs):
+        self.filter = _filter
+        self.func = lambda stream: stream.filter(_filter, **filter_kwargs)
+
+
+def build_effects_ffmpeg(audio_signal, filters, silent=False):
+    """
+    build_effects_ffmpeg takes an AudioSignal object and a list of FFmpegFilter objects
+    and sequentially applies each filter to the signal. 
+    Args:
+        audio_signal (AudioSignal): AudioSignal object
+        filters(list): List of FFmpegFilter objects
+        silent (bool): If True, suppresses all FFmpeg output. If False, FFmpeg will log information
+        with loglevel 'info'
+    Returns:
+        augmented_signal(AudioSignal): A new AudioSignal object, with the audio data from 
+        audio_signal after applying filters
+
+    """
+    # lazy load to avoid circular import
+    from . import AudioSignal
+
+    tmpfiles = []
+
+    input_args = {}
+    if silent:
+        input_args['loglevel'] = 'quiet'
+
+    with _close_temp_files(tmpfiles):
+        curr_tempfile = tempfile.NamedTemporaryFile(suffix=".flac")
+        out_tempfile = tempfile.NamedTemporaryFile(suffix=".flac")
+        
+        tmpfiles.append(curr_tempfile)
+        tmpfiles.append(out_tempfile)
+        audio_signal.write_audio_to_file(curr_tempfile)
+
+        tmpfiles.append(curr_tempfile)
+        stream = ffmpeg.input(curr_tempfile.name, **input_args)
+        for _filter in filters:
+            stream = _filter(stream)
+        (stream
+         .output(out_tempfile.name)
+         .overwrite_output()
+         .run()
+         )
+
+        augmented_signal = AudioSignal(path_to_input_file=out_tempfile.name)
```

Comment:
This won't keep metadata from the input `audio_signal`. I'm pretty sure the only metadata that you'll want to keep is `audio_signal.label`. This will also set `augmented_signal`'s path to the tempfile name, instead of the path of the input audio signal. I'm torn about whether this is okay or not. What are your thoughts @pseeth ?

Reply:
I'll copy over in the `audio_signal.label` to the new signal. I think it would make sense if the `augmented_signal`'s path is just set to `None`.

Reply:
You should also copy over effects metadata (https://github.com/nussl/nussl/pull/162#commitcomment-39648466)